lib/livebook/hubs/team_client.ex

```diff
@@ -534,4 +588,68 @@ defmodule Livebook.Hubs.TeamClient do
 
   defp nullify(""), do: nil
   defp nullify(value), do: value
+
+  defp deploy_apps(id, %{id: id} = deployment_group, derived_key) do
+    app_deployments =
+      for app_deployment <- deployment_group.app_deployments do
+        {:ok, app_deployment} = download_and_deploy(app_deployment, derived_key)
+        app_deployment
+      end
+
+    %{deployment_group | app_deployments: app_deployments}
+  end
+
+  defp deploy_apps(_, deployment_group, _), do: deployment_group
+
+  defp undeploy_apps(id, %{id: id} = deployment_group) do
+    deployment_group
+  end
+
+  defp undeploy_apps(_, deployment_group), do: deployment_group
+
+  defp download_and_deploy(%{file: nil} = app_deployment, _) do
+    app_deployment
+  end
+
+  defp download_and_deploy(app_deployment, derived_key) do
+    download_path = Path.join(System.tmp_dir!(), app_deployment.filename)
+    download_file = FileSystem.File.new(FileSystem.Local.new(), download_path)
+    destination_path = app_deployment_path(app_deployment)
+    archive_url = app_deployment.file
+    app_deployment = %{app_deployment | file: download_file}
+
+    with {:ok, %{status: 200} = response} <- Req.get(archive_url),
+         :ok <- FileSystem.File.write(download_file, response.body),
+         :ok <- undeploy_app(app_deployment),
+         :ok <- decrypt_zip(download_file, derived_key),
+         :ok <- Teams.AppDeployment.unzip_app(app_deployment, destination_path),
+         :ok <- Livebook.Apps.deploy_apps_in_dir(destination_path),
+         :ok <- FileSystem.File.remove(download_file) do
+      {:ok, %{app_deployment | file: nil}}
+    end
+  end
+
+  defp undeploy_app(app_deployment) do
+    with {:ok, app} <- Livebook.Apps.fetch_app(app_deployment.slug) do
+      Livebook.App.close(app.pid)
+    end
+
+    file_system = FileSystem.Local.new()
+    app_file = FileSystem.File.new(file_system, app_deployment_path(app_deployment))
+
+    case FileSystem.File.exists?(app_file) do
+      {:ok, true} -> FileSystem.File.remove(app_file)
+      {:ok, false} -> :ok
+    end
+  end
+
+  defp decrypt_zip(file, derived_key) do
+    with {:ok, encrypted_content} <- FileSystem.File.read(file),
+         {:ok, decrypted_content} <- Teams.decrypt(encrypted_content, derived_key) do
+      FileSystem.File.write(file, decrypted_content)
+    end
+  end
+
+  defp app_deployment_path(app_deployment),
+    do: Path.join(Livebook.Config.apps_path(), app_deployment.slug)
```

Comment:
We should probably have it in a temp directory instead with a random name. Something like:

    Path.join(Livebook.Config.tmp_path(), app_deployment.slug <> Base.url_encode64(:crypto.strong_rand_bytes(6)))

Reply:
Since it's an app deployment, wouldn't make sense to have it in the `apps_path/0`?